### PR TITLE
Optimize PromQL aggregations

### PR DESCRIPTION
**Pull request by free**
Compute hash of label subsets without creating a `labels.Labels` first. Doubles aggregation speed and reduces memory usage by ~10x for range queries with 1000+ steps.

Attached [before](https://github.com/prometheus/prometheus/files/2087282/before.txt) and [after](https://github.com/prometheus/prometheus/files/2087281/after.txt) benchmark results (compare the `sum()` benchmarks)

### CODE REVIEW

File: promql/engine.go [hunk 1]
```diff
@@ -1521,13 +1485,18 @@ func (ev *evaluator) aggregation(op ItemType, grouping []string, without bool, p
 			var m labels.Labels
 
 			if without {
-				m = metric
+				lb := labels.NewBuilder(metric)
+				if without {
```
[the remark this answers is not in the document]
Comment (free) [reply promql/engine.go]: Sorry, copy-pasta. Removed.

**Comment by free** [comment]
I've played a bit with replacing the keys of `promql.EvalNodeHelper.sigf` and `promql.EvalNodeHelper.dmn` caches (currently a hash of the labels) with a pointer to the first label and squeezed out another 10-20% on binary operations, relabels and label_* functions. (I would have used the pointer to `label.Labels` but I believe that's actually a reference and it is actually copied around, whereas the address of the first label is constant during the lifetime of a `promql.EvalNodeHelper`.)

Should I bundle it in this PR or create a new one?

**Comment by brian-brazil** [comment]
Separate PR is best if there's no overlap.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**